test/controllers/products_controller_test.rb

```diff
@@ -1,7 +1,167 @@
 require "test_helper"
 
 describe ProductsController do
-  # it "does a thing" do
-  #   value(1+1).must_equal 2
-  # end
+  before do
+    @user = User.create(username: "test user", email: "test_email@example.com") 
+    @category1 = Category.create(name: "flower")
+    @category2 = Category.create(name: "specialty")
+    @product = Product.create(user_id: @user.id, name: "test product", description: "cool product", price: 1.99, photo_url: "url", stock: 3)
+    
+    @review1 = Review.create(product_id: @product.id, rating: 5, comment: "great product")
+    @review2 = Review.create(product_id: @product.id, rating: 2, comment: "bad product")
+    
+    @order = Order.create(email: "test email", address: "test address", name: "test user name", cc_num: 1234, zip: 98111)
+    @order_item1 = OrderItem.create(product_id: @product.id, order_id: @order.id, quantity: 2)
+    @order_item2 = OrderItem.create(product_id: @product.id, order_id: @order.id, quantity: 4)
+  end
+  
+  # describe "Logged in users" do
+  #   # before do
+  #   #   perform_login(users(:grace))
+  #   # end
+  
+  describe "index" do
+    it "gives back success when products saved" do
+      get products_path
+      must_respond_with :success
+    end 
+    
+    it "can get the root path" do
+      get root_path
+      must_respond_with :success
+    end
+    
+    it "responds with sucess when no products saved" do
+      Product.destroy_all
+      
+      expect(Product.count).must_equal 0
+      get products_path
+      must_respond_with :success
+    end
+  end 
+  
+  describe "show" do
+    it "can get a valid product" do 
+      get product_path(@product.id)
+      must_respond_with :success
+    end 
+    
+    it "will redirect for an invalid product id" do
+      get product_path(-1111)
+      must_respond_with :redirect
+    end
+  end  
+  
+  describe "new" do 
+    it "can get the new product page" do
+      get new_product_path
+      must_respond_with :success
+    end
+  end
+  
+  describe "create" do
+    it "can create a new product with valid information" do 
+      product_hash = {
+      product: {
+      user_id: @user.id,
+      name: "new product",
+      description: "new description",
+      price: 1.99,
+      stock: 49,
+      photo_url: "new_photo url"
+    }
+  }
+  
+  expect {
+  post products_path, params: product_hash
+}.must_change "Product.count", 1
+
+new_product = Product.find_by(name: product_hash[:product][:name])  
+expect(new_product.price).must_equal product_hash[:product][:price]
+expect(new_product.stock).must_equal product_hash[:product][:stock]
+expect(new_product.photo_url).must_equal product_hash[:product][:photo_url]
+expect(new_product.description).must_equal product_hash[:product][:description]
+
+must_respond_with :redirect
+must_redirect_to product_path(new_product.id)
+end 
+end 
```

Comment:
Some of the indenting here is odd

Reply:
I'm sorry, I can't fix this indentation for anything on this planet :( 
I will try again when I refactor...